<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузить-и-изучить-данные" data-toc-modified-id="Загрузить-и-изучить-данные-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузить и изучить данные</a></span></li><li><span><a href="#Анализ-данных" data-toc-modified-id="Анализ-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Анализ данных</a></span><ul class="toc-item"><li><span><a href="#Посчитать,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитать,-сколько-книг-вышло-после-1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Посчитать, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Для каждой книги посчитать количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Определить издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Определить-автора-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Определить-автора-с-самой-высокой-средней-оценкой-книг-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Определить автора с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# АНАЛИЗ БАЗЫ ДАННЫХ ПРИ ПОМОЩИ SQL ЗАПРОСОВ

**Наша задача** Проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта

**Нам необходимо**

* Посчитать, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитать количество обзоров и среднюю оценку;
* Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
* Определить автора с самой высокой средней оценкой книг — учтем только книги с 50 и более оценками;
* Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.



**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<a id='start'></a>

**План работы**

1. <a href='#step_1'>Загрузить и изучить данные</a>  
2. <a href='#step_2'>Анализ данных</a>  
   2.1. <a href='#step_2.1'>Посчитать, сколько книг вышло после 1 января 2000 года</a>  
   2.2. <a href='#step_2.2'>Для каждой книги посчитать количество обзоров и среднюю оценку</a>      
   2.3. <a href='#step_2_3'>Определить издательство, которое выпустило наибольшее число книг толще 50 страниц</a>  
   2.4. <a href='#step_2_4'>Определить автора с самой высокой средней оценкой книг</a>  
   2.5. <a href='#step_2_5'>Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a>        
3. <a href='#step_3'>Общий вывод</a>    
   

<a id="step_1"></a>
## Загрузить и изучить данные

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [3]:
def table(tab):
    query = '''
    SELECT *
    FROM {}
    LIMIT 5
    '''.format(tab) 
    return pd.io.sql.read_sql(query, con = engine)

In [4]:
def my_function(query):
    return pd.io.sql.read_sql(query, con = engine)

In [5]:
table('books')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
query = '''
SELECT 
    COUNT(*)
FROM books
'''

my_function(query)

count
0   1000

В таблице books информация о 1000 книгах

In [7]:
table('authors')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
query = '''
SELECT 
    COUNT(*)
FROM authors
'''

my_function(query)

count
0    636

В таблице authors информация о 636 авторах

In [9]:
table('publishers')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [10]:
query = '''
SELECT 
    COUNT(*)
FROM publishers
'''

my_function(query)

count
0    340

В таблице publishers информация о 340 издательствах

In [11]:
table('ratings')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [12]:
query = '''
SELECT 
    COUNT(*)
FROM ratings
'''

my_function(query)

count
0   6456

В таблице ratings информация о 6456 оценках книг

In [13]:
table('reviews')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [14]:
query = '''
SELECT 
    COUNT(*)
FROM reviews
'''

my_function(query)

count
0   2793

В таблицу reviews информация о 2793 обзорах на книги.

<a href='#start'>Вернуться к оглавлению</a>

**Вывод**  
Мы подключились к базе данных, выгрузили 5 таблиц: 
* В books информация о 1000 книгах
* В authors информация о 636 авторах
* В publishers информация о 340 издательствах
* В ratings информация о 6456 оценках книг
* В reviews информация о 2793 обзорах на книги.

<a href='#start'>Вернуться к оглавлению</a>

In [15]:
def count(tab):
    query = '''
    SELECT COUNT(*)
    FROM {}
    '''.format(tab) 
    return pd.io.sql.read_sql(query, con = engine)

In [16]:
a = ['books','authors','publishers','ratings','reviews']
for i in a:
    display(f'Название таблицы {i}')
    display(table(i))
    display(count(i))

'Название таблицы books'

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

count
0   1000

'Название таблицы authors'

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

count
0    636

'Название таблицы publishers'

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

count
0    340

'Название таблицы ratings'

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

count
0   6456

'Название таблицы reviews'

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

count
0   2793

<a id="step_2"></a>
## Анализ данных

<a id="step_2.1"></a>
### Посчитать, сколько книг вышло после 1 января 2000 года

In [17]:
query = '''
SELECT 
    COUNT(book_id)
FROM books
WHERE publication_date >'01-01-2000'
'''

my_function(query)

count
0    819

**Вывод**  
После 1 января 2000 года вышло в свет 819 книг

<a href='#start'>Вернуться к оглавлению</a>

<a id="step_2.2"></a>
### Для каждой книги посчитать количество обзоров и среднюю оценку

In [18]:
query = '''
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT(rev.review_id)) AS count_reviews,
       ROUND(AVG(rating), 2) AS average_rating 
FROM books as b
LEFT JOIN reviews as rev ON b.book_id  = rev.book_id
LEFT JOIN ratings as rat ON b.book_id=rat.book_id
GROUP BY b.book_id
ORDER BY count_reviews  DESC, 
         average_rating DESC

'''

my_function(query)

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...   
2        299  Harry Potter and the Chamber of Secrets (Harry...   
3        656                                     The Book Thief   
4        734                                   The Glass Castle   
..       ...                                                ...   
995      191  Disney's Beauty and the Beast (A Little Golden...   
996      387                               Leonardo's Notebooks   
997      221                          Essential Tales and Poems   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
999      808                            The Natural Way to Draw   

     count_reviews  average_rating  
0                7            3.66  
1                6            4.41  
2                6            4.29  
3                6            4.26  
4                6            4.21  
..             ...             ...  
995              0            4.00  
996              0            4.00  
997              0            4.00  
998              0            3.67  
999              0            3.00  

[1000 rows x 4 columns]

Посмотрим книги с наибольшим рейтингом

In [19]:
query = '''
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT(rev.review_id)) AS count_reviews,
       ROUND(AVG(rating), 2) AS average_rating 
FROM books as b
LEFT JOIN reviews as rev ON b.book_id  = rev.book_id
LEFT JOIN ratings as rat ON b.book_id=rat.book_id
GROUP BY b.book_id
ORDER BY average_rating DESC, count_reviews  DESC
         

'''

my_function(query)

book_id                                              title  \
0         17                      A Dirty Job (Grim Reaper  #1)   
1        553            School's Out—Forever (Maximum Ride  #2)   
2        444       Moneyball: The Art of Winning an Unfair Game   
3        642                 The Big Bad Wolf  (Alex Cross  #9)   
4        967               Welcome to Temptation (Dempseys  #1)   
..       ...                                                ...   
995      915  The World Is Flat: A Brief History of the Twen...   
996      202                                      Drowning Ruth   
997      316                  His Excellency: George Washington   
998      371                                              Junky   
999      303                               Harvesting the Heart   

     count_reviews  average_rating  
0                4            5.00  
1                3            5.00  
2                3            5.00  
3                2            5.00  
4                2            5.00  
..             ...             ...  
995              3            2.25  
996              3            2.00  
997              2            2.00  
998              2            2.00  
999              2            1.50  

[1000 rows x 4 columns]

**Вывод**  
В данных 1000 книг, отсортированные по количеству обзоров и среднему рейтингу: 

* Книга с наибольшим рейтингом (среди средних оценок 5) - A Dirty Job (Grim Reaper #1) -  на нее написано 4 обзора
* Есть один лидер по количеству обзоров - книга с id 948 - Twilight (Twilight #1), на нее написано 7 обзоров, рейтинг 3,66
* Есть книги, к которым не было написано ни одного пользовательских обзора  


<a href='#start'>Вернуться к оглавлению</a>

<a id="step_2.3"></a>
### Определить издательство, которое выпустило наибольшее число книг толще 50 страниц

In [20]:
query = '''

SELECT publisher,
       COUNT(book_id) AS count_books
FROM publishers AS p
INNER JOIN books AS b ON p.publisher_id = b.publisher_id
WHERE num_pages > 50
GROUP BY publisher
ORDER BY count_books DESC
LIMIT 1
'''

my_function(query)

publisher  count_books
0  Penguin Books           42

**Вывод**  
Издательство, которое выпустило наибольшее число книг толще 50 страниц — Penguin Books

<a href='#start'>Вернуться к оглавлению</a>

<a id="step_2.4"></a>
### Определить автора с самой высокой средней оценкой книг

Учтем только книги с 50 и более оценками

**Вывод**  
Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré - Джоан Роулинг, известная всему миру автор бестселлеров про Гарри Поттера.  
Средняя оценка 4,29 среди книг, которым поставили 50 и более оценок

In [22]:
query = '''

SELECT a.author, ROUND(AVG(rat.rating),2) AS average_rating
FROM authors AS a
LEFT JOIN books AS b ON a.author_id=b.author_id
LEFT JOIN ratings AS rat ON b.book_id=rat.book_id
WHERE b.book_id in (
-- Найдем книги с 50 и более оценками                    
                    SELECT b.book_id
                    FROM books as b
                    LEFT JOIN ratings AS r ON r.book_id = b.book_id
                    GROUP BY b.book_id
                    HAVING COUNT(rating) >= 50)

GROUP BY a.author
ORDER BY AVG(rat.rating) DESC
LIMIT 1
'''

my_function(query)

author  average_rating
0  J.K. Rowling/Mary GrandPré            4.29

<a href='#start'>Вернуться к оглавлению</a>

<a id="step_2.5"></a>
### Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [23]:
query = '''
WITH
username_count as (
--определяем пользователей, которые поставили больше 50 оценок
SELECT username
FROM ratings
GROUP BY username
HAVING COUNT(*) > 50),

count_reviews as (
--считаем количество обзоров по каждому пользователю
SELECT username,
       COUNT(review_id) AS count_reviews
FROM reviews 
GROUP BY username)

SELECT FLOOR(AVG(count_reviews))
FROM username_count as u
LEFT JOIN count_reviews as c ON u.username = c.username

'''

my_function(query)

floor
0   24.0

**Вывод**  
Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24

<a href='#start'>Вернуться к оглавлению</a>

<a id="step_3_end"></a>
## Общий вывод

В ходе работы мы проанализировали базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

**Наша задача** Проанализировать базу данных. Чтобы помочь сформулировать ценностное предложение для нового продукта.

**Исходные данные**
Мы подключились к базе данных, выгрузили 5 таблиц:

* В books информация о 1000 книгах
* В authors информация о 636 авторах
* В publishers информация о 340 издательствах
* В ratings информация о 6456 оценках книг
* В reviews информация о 2793 обзорах на книги.

**Результат анализа**

* После 1 января 2000 года вышло в свет 819 книг
* В данных 1000 книг, отсортированные по количеству обзоров и среднему рейтингу:   

  * Книга с наибольшим рейтингом (среди средних оценок 5) - A Dirty Job (Grim Reaper #1) -  на нее написано 4 обзора
  * Есть один лидер по количеству обзоров - книга с id 948 - Twilight (Twilight #1), на нее написано 7 обзоров, рейтинг 3,66
  * Есть книги, к которым не было написано ни одного пользовательских обзора    
* Издательство, которое выпустило наибольшее число книг толще 50 страниц — Penguin Books
* Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré - Джоан Роулинг, известная всему миру автор бестселлеров про Гарри Поттера.  
  Средняя оценка 4,29 среди книг, которым поставили 50 и более оценок
* Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24

<a href='#start'>Вернуться к оглавлению</a>